Importing requered libraries and Dataset

In [1]:
import tensorflow as tf
import matplotlib.pyplot as plt
from tensorflow.keras import datasets, layers, models, losses
from sklearn.metrics import classification_report, f1_score
import numpy as np

Splitting the dataset into training and test dataset

In [2]:
(x_train,y_train),(x_test,y_test) = datasets.mnist.load_data()
x_train.shape

(60000, 28, 28)

Padding the dataset with zeroes and 8 bit values to make test and train dataset to recieve 32 by 32 images

In [3]:
x_train = tf.pad(x_train, [[0, 0], [2,2], [2,2]])/255
x_test = tf.pad(x_test, [[0, 0], [2,2], [2,2]])/255
x_train.shape

TensorShape([60000, 32, 32])

Expanding the tensor 

In [4]:
x_train = tf.expand_dims(x_train, axis=3, name=None)
x_test = tf.expand_dims(x_test, axis=3, name=None)
x_train.shape

TensorShape([60000, 32, 32, 1])

Introducing the validation set

In [5]:
x_val = x_train[-2000:,:,:,:] 
y_val = y_train[-2000:] 
x_train = x_train[:-2000,:,:,:] 
y_train = y_train[:-2000]

Compiling the Lenet model and using tanh activation function

In [6]:
model = models.Sequential()
model.add(layers.Conv2D(6, 5, activation='tanh', input_shape=x_train.shape[1:]))
model.add(layers.AveragePooling2D(2))
model.add(layers.Activation('sigmoid'))
model.add(layers.Conv2D(16, 5, activation='tanh'))
model.add(layers.AveragePooling2D(2))
model.add(layers.Activation('sigmoid'))
model.add(layers.Conv2D(120, 5, activation='tanh'))
model.add(layers.Flatten())
model.add(layers.Dense(84, activation='tanh'))
model.add(layers.Dense(10, activation='softmax'))
model.summary()

c:\Users\bhuva\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 28, 28, 6)      │           156 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ average_pooling2d               │ (None, 14, 14, 6)      │             0 │
│ (AveragePooling2D)              │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ activation (Activation)         │ (None, 14, 14, 6)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 10, 10, 16)     │         2,416 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ average_pooling2d_1             │ (None, 5, 5, 16)       │             0 │
│ (AveragePooling2D)              │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ activation_1 (Activation)       │ (None, 5, 5, 16)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 1, 1, 120)      │        48,120 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 120)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 84)             │        10,164 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │           850 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 61,706 (241.04 KB)

 Trainable params: 61,706 (241.04 KB)

 Non-trainable params: 0 (0.00 B)

Optimizing the model using Adam Optimizer and training the data till 40 epochs for accuracy

In [7]:
model.compile(optimizer='adam', loss=losses.sparse_categorical_crossentropy, metrics=['accuracy'])
history = model.fit(x_train, y_train, batch_size=64, epochs=40, validation_data=(x_val, y_val))

Epoch 1/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 7s 6ms/step - accuracy: 0.1650 - loss: 2.2008 - val_accuracy: 0.8790 - val_loss: 0.4493
Epoch 2/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.8377 - loss: 0.5091 - val_accuracy: 0.9345 - val_loss: 0.2392
Epoch 3/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.8925 - loss: 0.3408 - val_accuracy: 0.9490 - val_loss: 0.1981
Epoch 4/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.9136 - loss: 0.2773 - val_accuracy: 0.9500 - val_loss: 0.1762
Epoch 5/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.9261 - loss: 0.2344 - val_accuracy: 0.9550 - val_loss: 0.1732
Epoch 6/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 8s 8ms/step - accuracy: 0.9311 - loss: 0.2116 - val_accuracy: 0.9650 - val_loss: 0.1413
Epoch 7/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 8s 9ms/step - accuracy: 0.9384 - loss: 0.1990 - val_accuracy: 0.9690 - val_loss: 0.1142
Epoch 8/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 9s 10ms/step - accuracy: 0.9451 - loss: 0.1701 - val_accuracy: 0

Now testing the model with the test dataset

In [8]:
model.evaluate(x_test, y_test)
print(f"Accuracy on test set: {model.evaluate(x_test, y_test)[1]:.2f}")

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9780 - loss: 0.0763
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9780 - loss: 0.0763
Accuracy on test set: 0.98


Predicting labels on test set

In [9]:
y_pred_probs = model.predict(x_test)  
y_pred_classes = np.argmax(y_pred_probs, axis=1)  


313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step


Computing f1 Score
The f1 Score ranges from 1 to 0

In [10]:
f1 = f1_score(y_test, y_pred_classes, average='macro')
print(f"F1 Score (Weighted): {f1:.2f}")


F1 Score (Weighted): 0.98


After changing Number of Epochs:

    Accuracy : 0.98

    f1 Score : 0.98

After modifying the Epoch we got the above accuracy and f1 score as the result